include/seqan3/io/alignment_file/format_bam.hpp

```diff
@@ -0,0 +1,1061 @@
+            stream_it = '\0';
+
+            // write cigar
+            for (auto [cigar_op, cigar_count] : cigar)
+            {
+                cigar_count = cigar_count << 4;
+                cigar_count |= static_cast<int32_t>(char_to_sam_rank[cigar_op]);
+                std::ranges::copy_n(reinterpret_cast<char *>(&cigar_count), 4, stream_it);
+            }
+
+            // write seq (to sam_dna16 characters go into one byte)
+            using alph_t = value_type_t<seq_type>;
+            constexpr auto to_dna16 = detail::convert_through_char_representation<sam_dna16, alph_t>;
+
+            auto sit = std::ranges::begin(seq);
+            for (int32_t sidx = 0; sidx < ((core.l_seq & 1) ? core.l_seq - 1 : core.l_seq); ++sidx, ++sit)
+            {
+                uint8_t compressed_chr = to_rank(to_dna16[to_rank(*sit)]) << 4;
+                ++sidx, ++sit;
+                compressed_chr |= to_rank(to_dna16[to_rank(*sit)]);
+                stream_it = static_cast<char>(compressed_chr);
+            }
+
+            if (core.l_seq & 1) // write one more
+                stream_it = static_cast<char>(to_rank(to_dna16[to_rank(*sit)]) << 4);
+
+            // write qual
+            if (std::ranges::empty(qual))
+            {
+                auto v = view::repeat_n(static_cast<char>(255), core.l_seq);
+                std::ranges::copy_n(v.begin(), core.l_seq, stream_it);
+            }
+            else
+            {
+                assert(static_cast<int32_t>(std::ranges::distance(qual)) == core.l_seq);
+                auto v = qual | std::view::transform([] (auto chr) { return static_cast<char>(to_rank(chr)); });
+                std::ranges::copy_n(v.begin(), core.l_seq, stream_it);
+            }
+
+            // write optional fields
+            stream << tag_dict_binary_str;
+        } // if constexpr (!detail::decays_to_ignore_v<header_type>)
+    }
+
+    //!\brief Converts a cigar op character to the rank according to the official BAM specifications.
+    static constexpr std::array<uint8_t, 256> char_to_sam_rank
+    {
+        [] () constexpr
+        {
+            std::array<uint8_t, 256> ret{};
+
+            using index_t = std::make_unsigned_t<char>;
+
+            // ret['M'] = 0; set anyway by initialization
+            ret[static_cast<index_t>('I')] = 1;
+            ret[static_cast<index_t>('D')] = 2;
+            ret[static_cast<index_t>('N')] = 3;
+            ret[static_cast<index_t>('S')] = 4;
+            ret[static_cast<index_t>('H')] = 5;
+            ret[static_cast<index_t>('P')] = 6;
+            ret[static_cast<index_t>('=')] = 7;
+            ret[static_cast<index_t>('X')] = 8;
+
+            return ret;
+        }()
+    };
+
+    /*!\brief Writes the optional fields of the seqan3::sam_tag_dictionary.
+     * \param[in] tag_dict The tag dictionary to print.
+     */
+    auto get_tag_dict_str(sam_tag_dictionary const & tag_dict)
```

Comment:
```suggestion
    std::string get_tag_dict_str(sam_tag_dictionary const & tag_dict)
```
If you know the type, why not write it there?